numpyro/contrib/tfp/distributions.py

```diff
@@ -0,0 +1,182 @@
+        return self.bijector.forward(x)
+
+    def inv(self, y):
+        return self.bijector.inverse(y)
+
+    def log_abs_det_jacobian(self, x, y, intermediates=None):
+        return self.bijector.forward_log_det_jacobian(x, self.event_dim)
+
+
+@biject_to.register(BijectorConstraint)
+def _transform_to_bijector_constraint(constraint):
+    return BijectorTransform(constraint.bijector)
+
+
+class TFPDistributionMixin(NumPyroDistribution):
+    """
+    A mixin layer to make TensorFlow Probability (TFP) distribution compatible
+    with NumPyro internal.
+    """
+    def __call__(self, *args, **kwargs):
+        key = kwargs.pop('rng_key')
+        return self.sample(*args, seed=key, **kwargs)
+
+    @property
+    def support(self):
+        bijector = self._default_event_space_bijector()
+        if bijector is not None:
+            return BijectorConstraint(bijector)
+        else:
+            return None
+
+    @property
+    def is_discrete(self):
+        # XXX: this should cover most cases
+        return self.support is None
+
+
+class InverseGamma(tfd.InverseGamma):
+    arg_constraints = {"concentration": constraints.positive, "scale": constraints.positive}
+
+
+class OneHotCategorical(tfd.OneHotCategorical):
+    arg_constraints = {"logits": constraints.real_vector}
+    has_enumerate_support = True
+    support = constraints.simplex
+    is_discrete = True
+
+    def enumerate_support(self, expand=True):
+        n = self.event_shape[-1]
+        values = jnp.identity(n, dtype=canonicalize_dtype(self.dtype))
+        values = values.reshape((n,) + (1,) * len(self.batch_shape) + (n,))
+        if expand:
+            values = jnp.broadcast_to(values, (n,) + self.batch_shape + (n,))
+        return values
+
+
+class OrderedLogistic(tfd.OrderedLogistic):
+    arg_constraints = {"cutpoints": constraints.ordered_vector, "loc": constraints.real}
+
+
+class Pareto(tfd.Pareto):
+    arg_constraints = {"concentration": constraints.positive, "scale": constraints.positive}
+
+
+__all__ = ['BijectorConstraint', 'BijectorTransform', 'TFPDistributionMixin']
+_len_all = len(__all__)
+for _name, _Dist in tfd.__dict__.items():
+    if not isinstance(_Dist, type):
+        continue
+    if not issubclass(_Dist, tfd.Distribution):
+        continue
+    if _Dist is tfd.Distribution:
+        continue
+
+    try:
+        _PyroDist = locals()[_name]
+    except KeyError:
+        _PyroDist = type(_name, (_Dist, TFPDistributionMixin), {})
+        _PyroDist.__module__ = __name__
+        if hasattr(numpyro_dist, _name):
+            numpyro_dist_class = getattr(numpyro_dist, _name)
+            # resolve FooProbs/FooLogits namespaces
+            if type(numpyro_dist_class).__name__ == "function":
+                if not hasattr(numpyro_dist, _name + "Logits"):
+                    continue
+                numpyro_dist_class = getattr(numpyro_dist, _name + "Logits")
+            _PyroDist.arg_constraints = numpyro_dist_class.arg_constraints
```

Comment:
I think this is fine. Tensorflow has its own mechanism to validate args. I copied `arg_constraints` to TFP distribution so that users can use some TFP distributions in `scan` (this is actually unnecessary because we already have them implemented in NumPyro). I don't intend to support `scan` for all TFP distributions though (because I don't understand how some special TFP distributions, such as JointDistribution, work and some TFP distributions do not take `ndarray` as its args...).

Reply:
I see. My comment was motivated by the eventual goal of replacing existing NumPyro distributions and transforms with their TFP counterparts in NumPyro and Funsor. If `arg_constraints` are not necessary to achieve that (and they shouldn't be, assuming we can implement `funsor.jax.distributions.Distribution._infer_param_domain` directly without relying on `arg_constraints`), then I agree it's fine.